File: app/src/main/java/com/helenacorp/android/mybibliotheque/SignupActivity.java
package com.helenacorp.android.mybibliotheque;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.android.material.textfield.TextInputEditText;
import com.google.android.material.textfield.TextInputLayout;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.auth.UserProfileChangeRequest;
import com.google.firebase.database.DatabaseReference;
import com.helenacorp.android.mybibliotheque.Controllers.Activities.AccountActivity;

import java.util.regex.Pattern;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;
import androidx.appcompat.app.AppCompatActivity;

public class SignupActivity extends AppCompatActivity implements View.OnClickListener {
    private static final String EMAIL_PATTERN = "^[a-zA-Z0-9#_~!$&'()*+,;=:.\"(),:;<>@\\[\\]\\\\]+@[a-zA-Z0-9-]+(\\.[a-zA-Z0-9-]+)*$";
    private Pattern pattern = Pattern.compile(EMAIL_PATTERN);

    private FirebaseAuth mAuth;
    private FirebaseUser user;
    private DatabaseReference ref;
    private FirebaseAuth.AuthStateListener mAuthListener;
    private TextInputEditText authName, authEmail, authPassw;
    private TextInputLayout authNameParent, authEmailParent, authPassParent;
    private Button btnCreatAccount;
    private String displayName, useremail, password, name;
    private View viewLayout;
    private TextView messToast, txtTitle;

    @RequiresApi(api = Build.VERSION_CODES.O)
    @SuppressLint("WrongViewCast")
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_signup);

        LayoutInflater layoutInflater = getLayoutInflater();
        viewLayout = layoutInflater.inflate(R.layout.custom_toast, (ViewGroup) findViewById(R.id.custom_layout));
        messToast = (TextView) viewLayout.findViewById(R.id.toast_txt);
        txtTitle = findViewById(R.id.txt_signup);
        authName = findViewById(R.id.name_signup);
        authEmail = findViewById(R.id.email_signup);
        authPassw = findViewById(R.id.password_signup);
        authNameParent = findViewById(R.id.name_singup_parent);
        authEmailParent = findViewById(R.id.email_signup_parent);
        authPassParent = findViewById(R.id.password_signup_parent);

        btnCreatAccount = (Button) findViewById(R.id.btn_creatAccount);
        btnCreatAccount.setOnClickListener(this);
        authName.addTextChangedListener(new MyTextWatcher(authName));
        authEmail.addTextChangedListener(new MyTextWatcher(authEmail));
        authPassw.addTextChangedListener(new MyTextWatcher(authPassw));

        mAuth = FirebaseAuth.getInstance();
    }

    public void validation() {
        if (!validateName()) {
            return;
        }

        if (!validateEmail()) {
            return;
        }

        if (!validatePassword()) {
            return;
        }
        loginAccount(displayName = authName.getText().toString(), authEmail.getText().toString(), authPassw.getText().toString());
    }

    private void loginAccount(final String name, String email, String password) {
        // START create_user_with_email
        mAuth.createUserWithEmailAndPassword(email, password)
                .addOnCompleteListener(this, new OnCompleteListener<AuthResult>() {
                    @Override
                    public void onComplete(@NonNull Task<AuthResult> task) {
                        if (task.isSuccessful()) {
                            // Sign in success, update UI with the signed-in user's information
                            messToast.setText(R.string.mlog_bvn);
                            messageToast();
                            user = mAuth.getCurrentUser();
                            UserProfileChangeRequest profileUpdates = new UserProfileChangeRequest.Builder()
                                    .setDisplayName(displayName)
                                    .build();
                            user.updateProfile(profileUpdates).addOnCompleteListener(new OnCompleteListener<Void>() {
                                @Override
                                public void onComplete(@NonNull Task<Void> task) {
                                    messToast.setText(R.string.mlog_bvn);
                                    messageToast();
                                    Intent intent = new Intent(SignupActivity.this, AccountActivity.class);
                                    finish();
                                    startActivity(intent);
                                }
                            });


                        } else {
                            messToast.setText(R.string.mlog_signup);
                            messageToast();
                            // If sign in fails, display a message to the user.
                        }

                    }

                });
    }

    public void messageToast() {
        Toast toast1 = Toast.makeText(SignupActivity.this, " ", Toast.LENGTH_SHORT);
        toast1.setGravity(Gravity.CENTER, 0, 0);
        toast1.setView(viewLayout);
        toast1.show();
    }

    private boolean validateName() {
        if (authName.getText().toString().trim().length() < 3) {
            authNameParent.setError(getString(R.string.error_name));
            requestFocus(authName);
            return false;
        } else {
            authNameParent.setErrorEnabled(false);
        }

        return true;
    }

    private boolean validatePassword() {
        if (authPassw.getText().toString().trim().length() < 6) {
            authPassParent.setError(getString(R.string.error_password));
            requestFocus(authPassw);
            return false;
        } else {
            authPassParent.setErrorEnabled(false);
        }

        return true;
    }

    public boolean validateEmail() {
        String email = authEmail.getText().toString().trim();
        if (email.isEmpty() || !isValidEmail(email)) {
            authEmailParent.setError(getString(R.string.error_email));
            requestFocus(authEmail);
            return false;
        } else {
            authEmailParent.setErrorEnabled(false);
        }
        return true;
    }

    private static boolean isValidEmail(String email) {
        return !TextUtils.isEmpty(email) && android.util.Patterns.EMAIL_ADDRESS.matcher(email).matches();
    }

    private void requestFocus(View view) {
        if (view.requestFocus()) {
            getWindow().setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_VISIBLE);
        }
    }


/*
    @Override
    public void onStop(){
        super.onStop();
        if (mAuthListener != null) {
            mAuth.removeAuthStateListener(mAuthListener);
        }
    }
    @Override
    public void onStart(){
        super.onStart();
        mAuth.addAuthStateListener(mAuthListener);
    }*/

    @Override
    public void onClick(View view) {
        if (view == btnCreatAccount) {
            validation();
        }
    }

    private class MyTextWatcher implements TextWatcher {

        private View view;

        private MyTextWatcher(View view) {
            this.view = view;
        }

        public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {
        }

        public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
        }

        public void afterTextChanged(Editable editable) {
            switch (view.getId()) {
                case R.id.name_signup:
                    validateName();
                    break;
                case R.id.email_signup:
                    validateEmail();
                    break;
                case R.id.password_signup:
                    validatePassword();
                    break;
            }
        }
    }
}

File: app/src/main/java/com/helenacorp/android/mybibliotheque/SignupActivity.kt
package com.helenacorp.android.mybibliotheque

import android.annotation.SuppressLint
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.text.Editable
import android.text.TextUtils
import android.text.TextWatcher
import android.util.Patterns
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.view.WindowManager
import android.widget.Button
import android.widget.TextView
import android.widget.Toast
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AppCompatActivity
import com.google.android.material.textfield.TextInputEditText
import com.google.android.material.textfield.TextInputLayout
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseAuth.AuthStateListener
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.auth.UserProfileChangeRequest
import com.google.firebase.database.DatabaseReference
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.ktx.firestore
import com.google.firebase.ktx.Firebase
import com.helenacorp.android.mybibliotheque.Controllers.Activities.AccountActivity
import com.helenacorp.android.mybibliotheque.model.User
import java.util.regex.Pattern

class SignupActivity : AppCompatActivity(), View.OnClickListener {
    private val pattern = Pattern.compile(EMAIL_PATTERN)
    private var mAuth: FirebaseAuth? = null
    private var user: FirebaseUser? = null
    private val ref: DatabaseReference? = null
    private val mAuthListener: AuthStateListener? = null
    private var authName: TextInputEditText? = null
    private var authEmail: TextInputEditText? = null
    private var authPassw: TextInputEditText? = null
    private lateinit var mFirestore:FirebaseFirestore
    private var authNameParent: TextInputLayout? = null
    private var authEmailParent: TextInputLayout? = null
    private var authPassParent: TextInputLayout? = null
    private var btnCreatAccount: Button? = null
    private var displayName: String? = null
    private val useremail: String? = null
    private val password: String? = null
    private val name: String? = null
    private var viewLayout: View? = null
    private var messToast: TextView? = null
    private var txtTitle: TextView? = null

    @SuppressLint("WrongViewCast")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_signup)
        val layoutInflater = layoutInflater
        val viewgroup = findViewById<ViewGroup>(R.id.custom_layout)
        viewLayout = layoutInflater.inflate(R.layout.custom_toast, viewgroup)
        messToast = viewLayout!!.findViewById<View>(R.id.toast_txt) as TextView
        txtTitle = findViewById(R.id.txt_signup)
        authName = findViewById(R.id.name_signup)
        authEmail = findViewById(R.id.email_signup)
        authPassw = findViewById(R.id.password_signup)
        authNameParent = findViewById(R.id.name_singup_parent)
        authEmailParent = findViewById(R.id.email_signup_parent)
        authPassParent = findViewById(R.id.password_signup_parent)
        btnCreatAccount = findViewById<View>(R.id.btn_creatAccount) as Button
        btnCreatAccount!!.setOnClickListener(this)
        authName!!.addTextChangedListener(MyTextWatcher(authName))
        authEmail!!.addTextChangedListener(MyTextWatcher(authEmail))
        authPassw!!.addTextChangedListener(MyTextWatcher(authPassw))
        mAuth = FirebaseAuth.getInstance()
        mFirestore = Firebase.firestore
    }

    fun validation() {
        if (!validateName()) {
            return
        }
        if (!validateEmail()) {
            return
        }
        if (!validatePassword()) {
            return
        }
        loginAccount(authName!!.text.toString().also { displayName = it }, authEmail!!.text.toString(), authPassw!!.text.toString())
    }

    private fun loginAccount(name: String, email: String, password: String) {
        // START create_user_with_email
        mAuth!!.createUserWithEmailAndPassword(email, password)
                .addOnCompleteListener(this) { task ->
                    if (task.isSuccessful) {
                        // Sign in success, update UI with the signed-in user's information
                        messToast!!.setText(R.string.mlog_bvn)
                        messageToast()
                        user = mAuth!!.currentUser
                        val users = User(user!!.uid,displayName, email,null,null)
                        mFirestore.collection("users").document().set(users)
                        val profileUpdates = UserProfileChangeRequest.Builder()
                                .setDisplayName(displayName)
                                .build()
                        user!!.updateProfile(profileUpdates).addOnCompleteListener {
                            messToast!!.setText(R.string.mlog_bvn)
                            messageToast()
                            val intent = Intent(this@SignupActivity, AccountActivity::class.java)
                            finish()
                            startActivity(intent)
                        }
                    } else {
                        messToast!!.setText(R.string.mlog_signup)
                        messageToast()
                        // If sign in fails, display a message to the user.
                    }
                }
    }

    fun messageToast() {
        val toast1 = Toast.makeText(this@SignupActivity, " ", Toast.LENGTH_SHORT)
        toast1.setGravity(Gravity.CENTER, 0, 0)
        toast1.view = viewLayout
        toast1.show()
    }

    private fun validateName(): Boolean {
        if (authName!!.text.toString().trim { it <= ' ' }.length < 3) {
            authNameParent!!.error = getString(R.string.error_name)
            requestFocus(authName)
            return false
        } else {
            authNameParent!!.isErrorEnabled = false
        }
        return true
    }

    private fun validatePassword(): Boolean {
        if (authPassw!!.text.toString().trim { it <= ' ' }.length < 6) {
            authPassParent!!.error = getString(R.string.error_password)
            requestFocus(authPassw)
            return false
        } else {
            authPassParent!!.isErrorEnabled = false
        }
        return true
    }

    fun validateEmail(): Boolean {
        val email = authEmail!!.text.toString().trim { it <= ' ' }
        if (email.isEmpty() || !isValidEmail(email)) {
            authEmailParent!!.error = getString(R.string.error_email)
            requestFocus(authEmail)
            return false
        } else {
            authEmailParent!!.isErrorEnabled = false
        }
        return true
    }

    private fun requestFocus(view: View?) {
        if (view!!.requestFocus()) {
            window.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_VISIBLE)
        }
    }

    /*
    @Override
    public void onStop(){
        super.onStop();
        if (mAuthListener != null) {
            mAuth.removeAuthStateListener(mAuthListener);
        }
    }
    @Override
    public void onStart(){
        super.onStart();
        mAuth.addAuthStateListener(mAuthListener);
    }*/
    override fun onClick(view: View) {
        if (view === btnCreatAccount) {
            validation()
        }
    }

    private inner class MyTextWatcher(private val view: View?) : TextWatcher {
        override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
        override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
        override fun afterTextChanged(editable: Editable) {
            when (view!!.id) {
                R.id.name_signup -> validateName()
                R.id.email_signup -> validateEmail()
                R.id.password_signup -> validatePassword()
            }
        }

    }

    companion object {
        private const val EMAIL_PATTERN = "^[a-zA-Z0-9#_~!$&'()*+,;=:.\"(),:;<>@\\[\\]\\\\]+@[a-zA-Z0-9-]+(\\.[a-zA-Z0-9-]+)*$"
        private fun isValidEmail(email: String): Boolean {
            return !TextUtils.isEmpty(email) && Patterns.EMAIL_ADDRESS.matcher(email).matches()
        }
    }
}